lib/engine/game/g_18_norway/steps/token.rb

```diff
@@ -9,37 +9,44 @@ module Step
         class Token < Engine::Step::Token
           def available_hex(entity, hex)
             return true if super(entity, hex)
-            return true if @game.ferry_graph.reachable_hexes(entity)[hex]
+
+            return true if harbor_available?(entity, hex)
 
             false
           end
 
+          def harbor_available?(entity, hex)
+            city = @game.harbor_city_coordinates(hex.id)
+            return false if city.nil?
+
+            other_hex = @game.hex_by_id(city)
+
+            @game.ferry_graph.reachable_hexes(entity)[other_hex] ||
+            @game.token_graph_for_entity(entity).reachable_hexes(entity)[other_hex]
+          end
+
           def can_place_token?(_entity)
             true
           end
 
           def process_place_token(action)
             entity = action.entity
             city = action.city
-            connected_city = @game.loading || @game.token_graph_for_entity(entity).connected_nodes(entity)[city]
-            place_token(entity, action.city, action.token) if connected_city
-
-            unless connected_city
+            connected_city = @game.token_graph_for_entity(entity).connected_nodes(entity)[city]
+            if connected_city
+              place_token(entity, action.city, action.token)
+            elsif harbor_available?(entity, city.hex)
               @game.abilities(entity, :token) do |ability|
                 place_token(entity, action.city, action.token, connected: false, special_ability: ability)
+                entity.add_ability(Ability::Token.new(type: 'token', hexes: Engine::Game::G18Norway::Game::HARBOR_HEXES,
+                                                      from_owner: true, discount: 0, connected: true))
```

Comment:
Instead of adding a new token ability, could you give the original one a `count` attribute that's large enough for it to be reused throughout the game?

Reply:
This is not trivial since tokener.rb does not use the ability but remove it instead.

I have tried to change it but am afraid on how that will affect other games that uses ability for token
I have seen a problem with 1822CA and P10 which then would need an update to special_token
Would it be ok if I did this in another commit, want to make sure that I do not affect any other game to solve this

Reply:
Yes, putting this in a separate commit would be sensible.